crates/bevy_reflect/src/reflect.rs

```diff
@@ -34,27 +46,98 @@ pub enum ReflectMut<'a> {
 /// value passed in. If this is not done, [`Reflect::downcast`](trait.Reflect.html#method.downcast)
 /// will be UB (and also just logically broken).
 pub unsafe trait Reflect: Any + Send + Sync {
+    /// Returns the [type name] of the underlying type.
+    ///
+    /// [type name]: std::any::type_name
     fn type_name(&self) -> &str;
+
+    /// Returns the value as a [`&dyn Any`][std::any::Any].
     fn any(&self) -> &dyn Any;
+
+    /// Returns the value as a [`&mut dyn Any`][std::any::Any].
     fn any_mut(&mut self) -> &mut dyn Any;
+
+    /// Applies a reflected value to this value.
+    ///
+    /// When `Reflect` is derived for a type, the semantics of this method
+    /// depend on the kind of type `T`:
+    /// - If `T` is a struct with named fields, then the value of each named
+    ///   field of `value` is applied to the corresponding named field of
+    ///   `self`. Fields which are not present in both values are ignored.
+    /// - If `T` is a tuple struct, then the value of each numbered field is
+    ///   applied to the corresponding numbered field of `self.` Fields which
+    ///   are not present in both values are ignored.
+    /// - If `T` is a value type (i.e., neither of the above), then `value` is
+    ///   downcast to `T`, cloned, and assigned to `self`.
+    ///
+    /// Note that deriving `Reflect` for collections may result in undesired
+    /// behavior, such as applying not only the elements of `value` to those of
+    /// the other but also `value`'s metadata. For collection semantics,
+    /// consider manually implementing `Reflect` and using a helper function
+    /// such as [`list_apply`] to implement this method.
```

Comment:
I honestly don't know what happens when you try to derive Reflect on collections, I'll try it out. What metadata are you referring to?

I think we can detail the behaviour of tuples, maps and lists here too, so that external implementations have to conform with what we are doing for `Vec` and `HashMap`. (We should also change what HashMap does, but that's for another PR)

Reply:
By metadata I mean values that are part of the type, but aren't elements of the collection. Say a user has their own keyed `HashMap` type that they `#[derive(Reflect)]` on -- a derived implementation would ostensibly not apply only the entries, but also the hash key of `value`, which would produce undesired effects.